internal/jimm/applicationoffer.go

```diff
@@ -599,193 +546,57 @@ func (j *JIMM) FindApplicationOffers(ctx context.Context, user *openfga.User, fi
 		return nil, errors.E(op, errors.CodeBadRequest, "at least one filter must be specified")
 	}
 
-	offerFilters, err := j.applicationOfferFilters(ctx, filters...)
-	if err != nil {
-		return nil, errors.E(op, err)
-	}
-	userOfferUUIDs, err := user.ListApplicationOffers(ctx, ofganames.ReaderRelation)
-	if err != nil {
-		return nil, errors.E(op, err)
-	}
-	offerFilters = append(offerFilters, db.ApplicationOfferFilterByUUID(userOfferUUIDs))
-	offers, err := j.Database.FindApplicationOffers(ctx, offerFilters...)
-	if err != nil {
-		return nil, errors.E(op, err)
-	}
-	offerDetails := make([]jujuparams.ApplicationOfferAdminDetailsV5, len(offers))
-	for i, offer := range offers {
-		// TODO (alesstimec) Optimize this: currently check all possible
-		// permission levels for an offer, this is suboptimal.
-		accessLevel, err := j.getUserOfferAccess(ctx, user, &offer)
+	offers := []jujuparams.ApplicationOfferAdminDetailsV5{}
+	err := j.Database.ForEachController(ctx, func(ctl *dbmodel.Controller) error {
+		api, err := j.dial(ctx, ctl, names.ModelTag{})
 		if err != nil {
-			return nil, errors.E(op, err)
-		}
-
-		offerDetails[i] = offer.ToJujuApplicationOfferDetailsV5()
-
-		// non-admin users should not see connections of an application
-		// offer.
-		if accessLevel != "admin" {
-			offerDetails[i].Connections = nil
+			return errors.E(op, err)
 		}
-		users, err := j.listApplicationOfferUsers(ctx, offer.ResourceTag(), user.Identity, accessLevel)
+		defer api.Close()
+		controllerOffers, err := api.ListApplicationOffers(ctx, filters)
```

Comment:
I like this change of just querying the Juju controllers but we have some challenges with this approach:
1. It will query all offers on all controllers each time. That could be a large number of offers.
2. It doesn't take into account the fact that model owners have changed when a model was imported into JIMM.
3. It doesn't take into account the difference between `juju find-offers` and `juju offers`. The latter shows you the offers available in a model. The former shows you offers you have access to from all models so that you can use them in your current model.

I think we can solve this pretty easily with some code that already existed. List-offers should just list the offers in the specified model as was already done, no problems there. Find-offers can loop through all controllers/offers but requires the model owner to be treated carefully. If the model owner/name are specified we can just use the approach in list-offers.

Alternatively, we can consider removing the transfer of ownership concept when importing models into JIMM. That would mean two models with the same name/owner on different controllers can't be imported into JIMM but that could possibly be an okay tradeoff.

Reply:
@kian99 updated as you suggested:
- ListOffers now works as before
- FindApplicationOffers will query all controllers in parallel - not the most efficient, but better than the serial approach
- still does not handle the case of imported models